mssql/compiler.py

```diff
@@ -470,11 +480,43 @@ def as_sql(self):
 
         placeholder_rows, param_rows = self.assemble_as_sql(fields, value_rows)
 
-        if self.get_returned_fields() and self.connection.features.can_return_id_from_insert:
-            result.insert(0, 'SET NOCOUNT ON')
-            result.append((values_format + ';') % ', '.join(placeholder_rows[0]))
-            params = [param_rows[0]]
-            result.append('SELECT CAST(SCOPE_IDENTITY() AS bigint)')
+        if self.get_returned_fields() and self.can_return_columns_from_insert():
+            if self.can_return_rows_from_bulk_insert():
+                if not(self.query.fields):
+                    # There isn't really a single statement to bulk multiple DEFAULT VALUES insertions,
+                    # so we have to use a workaround:
+                    # https://dba.stackexchange.com/questions/254771/insert-multiple-rows-into-a-table-with-only-an-identity-column
+                    result = ["""
+                    WITH empty_rows AS (
```

Comment:
As far as I'm concerned, it's definitely a nicer workaround than the prev one (cf. prev comments), no contest.
Maybe just add some context here about how the trick works (8^4 = 4096, vs. 1000 (is it 1000?) expected as max bulk objs number, etc.)

Reply:
According to bulk_batch_size in operations.py, the bulk inserts are set to 1000 at a time (in batch).  The actual query can produce 4096 numbers in the sequence.  We could also limit it to **1024** (improving performance by some ms) by changing it to something corresponding to:
`WITH empty_rows AS (
    SELECT 1 AS n UNION ALL SELECT 1 AS n UNION ALL SELECT 1 AS n UNION ALL SELECT 1 AS n
)
SELECT row_number() over (order by a.n) as number
FROM empty_rows as a, empty_rows as b, empty_rows as c, empty_rows as d, empty_rows as e`